File: p11-platform/services/data-engine/connectors/entity_fallback.py
"""Deterministic GEO entity backfill when the analyzer returns an empty list."""
from __future__ import annotations

import re
from typing import Any, Dict, Iterable, List, Optional


GENERIC_WORDS = {'apartments', 'apartment', 'properties', 'property', 'living', 'homes'}


def find_mention_index(text: str, name: str) -> Optional[int]:
    trimmed = (name or '').strip()
    if not trimmed or len(trimmed) < 2:
        return None
    match = re.search(rf'\b{re.escape(trimmed)}\b', text or '', flags=re.IGNORECASE)
    return match.start() if match else None


def _normalize_domain(domain: str) -> str:
    normalized = (domain or '').lower().strip()
    normalized = re.sub(r'^https?://', '', normalized)
    normalized = re.sub(r'^www\.', '', normalized)
    return normalized.split('/')[0]


def _is_brand_name(entity_name: str, brand_name: str) -> bool:
    name = (entity_name or '').lower()
    brand = (brand_name or '').lower().strip()
    if not brand:
        return False
    if brand in name:
        return True
    words = [word for word in brand.split() if len(word) > 3]
    if not words:
        return False
    main = words[0]
    return len(main) >= 4 and main not in GENERIC_WORDS and main in name


def _is_brand_domain(domain: str, brand_domains: Iterable[str]) -> bool:
    normalized = _normalize_domain(domain)
    if not normalized:
        return False
    for candidate in brand_domains or []:
        brand = _normalize_domain(candidate)
        if brand and (normalized == brand or normalized.endswith('.' + brand)):
            return True
    return False


def find_tracked_brand_position(
    entities: Iterable[Dict[str, Any]],
    brand_name: str,
    brand_domains: Optional[Iterable[str]] = None,
) -> Optional[int]:
    domains = list(brand_domains or [])
    for entity in entities or []:
        if not isinstance(entity, dict):
            continue
        if domains and _is_brand_domain(entity.get('domain') or '', domains):
            return entity.get('position')
        if _is_brand_name(entity.get('name') or '', brand_name):
            return entity.get('position')
    return None


def _normalize_entities(entities: Optional[Iterable[Any]]) -> List[Dict[str, Any]]:
    seen = set()
    normalized: List[Dict[str, Any]] = []
    for entity in entities or []:
        if not isinstance(entity, dict):
            continue
        name = str(entity.get('name') or '').strip()
        if not name:
            continue
        key = name.lower()
        if key in seen:
            continue
        seen.add(key)
        position = entity.get('position')
        normalized.append({
            'name': name,
            'domain': str(entity.get('domain') or '').strip(),
            'rationale': str(entity.get('rationale') or 'Extracted from the answer text').strip(),
            'position': position if isinstance(position, int) and position > 0 else len(normalized) + 1,
        })
    normalized.sort(key=lambda item: item['position'])
    return normalized


def _known_names(brand_name: str, competitors: Optional[Iterable[str]] = None) -> List[str]:
    names: List[str] = []
    seen = set()
    for value in [brand_name, *(competitors or [])]:
        cleaned = str(value or '').strip()
        if not cleaned or '.' in cleaned or len(cleaned) < 2:
            continue
        key = cleaned.lower()
        if key in seen:
            continue
        seen.add(key)
        names.append(cleaned)
    return names


def _extract_entities_from_text(
    text: str,
    brand_name: str,
    brand_domains: Optional[Iterable[str]] = None,
    competitors: Optional[Iterable[str]] = None,
) -> List[Dict[str, Any]]:
    hits = []
    for name in _known_names(brand_name, competitors):
        index = find_mention_index(text, name)
        if index is None:
            continue
        hits.append((index, name))
    hits.sort(key=lambda item: item[0])
    domains = list(brand_domains or [])
    return [
        {
            'name': name,
            'domain': domains[0] if _is_brand_name(name, brand_name) and domains else '',
            'rationale': f'First mentioned in the answer at character {index + 1}.',
            'position': position,
        }
        for position, (index, name) in enumerate(hits, start=1)
    ]


def _ensure_brand_in_list(
    entities: List[Dict[str, Any]],
    text: str,
    brand_name: str,
    brand_domains: Optional[Iterable[str]] = None,
    competitors: Optional[Iterable[str]] = None,
) -> List[Dict[str, Any]]:
    if find_tracked_brand_position(entities, brand_name, brand_domains) is not None:
        return entities
    if find_mention_index(text, brand_name) is None:
        return entities
    extracted = _extract_entities_from_text(text, brand_name, brand_domains, competitors)
    brand = next((entity for entity in extracted if _is_brand_name(entity['name'], brand_name)), None)
    next_position = max((entity.get('position') or 0 for entity in entities), default=0) + 1
    return _normalize_entities([
        *entities,
        {
            'name': (brand or {}).get('name') or brand_name,
            'domain': (brand or {}).get('domain') or (list(brand_domains or []) or [''])[0],
            'rationale': (brand or {}).get('rationale') or 'Tracked brand was named in the answer but omitted from the extracted list.',
            'position': next_position,
        },
    ])


def ensure_ordered_entities(
    existing: Optional[Iterable[Any]],
    analysis_entities: Optional[Iterable[Any]],
    brand_name: str,
    brand_domains: Optional[Iterable[str]] = None,
    competitors: Optional[Iterable[str]] = None,
    text: str = '',
) -> List[Dict[str, Any]]:
    current = _normalize_entities(existing)
    if current:
        return _ensure_brand_in_list(current, text, brand_name, brand_domains, competitors)
    from_analysis = _normalize_entities(analysis_entities)
    if from_analysis:
        return _ensure_brand_in_list(from_analysis, text, brand_name, brand_domains, competitors)
    return _extract_entities_from_text(text, brand_name, brand_domains, competitors)


def reconcile_hallucination_flags(flags: Optional[Iterable[str]], text: str, brand_name: str) -> List[str]:
    next_flags = [flag for flag in (flags or [])]
    if find_mention_index(text, brand_name) is None:
        return next_flags
    return [flag for flag in next_flags if flag != 'possible_hallucination']


def finalize_answer_block(
    answer_block: Dict[str, Any],
    context: Dict[str, Any],
    natural_text: str = '',
    analysis: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    source_text = '\n'.join(
        part for part in [
            natural_text,
            context.get('sourceText') or '',
            answer_block.get('answer_summary') or '',
        ] if part
    )
    expected_city = context.get('expectedCity') or (context.get('propertyLocation') or {}).get('city')
    analysis_entities = (analysis or {}).get('ordered_entities') if isinstance(analysis, dict) else None
    answer_block['ordered_entities'] = ensure_ordered_entities(
        existing=answer_block.get('ordered_entities'),
        analysis_entities=analysis_entities,
        brand_name=context.get('brandName') or '',
        brand_domains=context.get('brandDomains') or [],
        competitors=context.get('competitors') or [],
        text=source_text,
    )
    notes = answer_block.get('notes') or {}
    notes['flags'] = reconcile_hallucination_flags(
        notes.get('flags') or [],
        source_text,
        context.get('brandName') or '',
    )
    answer_block['notes'] = notes
    if expected_city:
        answer_block.setdefault('_expected_city', expected_city)
    return answer_block

File: p11-platform/apps/web/utils/propertyaudit/entity-fallback.ts
export type FallbackEntity = {
  name: string
  domain: string
  rationale: string
  position: number
}

export type EntityFallbackContext = {
  brandName: string
  brandDomains?: string[]
  competitors?: string[]
  sourceText?: string
  expectedCity?: string
  analysisEntities?: Array<Partial<FallbackEntity> | null | undefined>
}

const GENERIC_WORDS = new Set(['apartments', 'apartment', 'properties', 'property', 'living', 'homes'])

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

export function findMentionIndex(text: string, name: string): number | null {
  const trimmed = name.trim()
  if (!trimmed || trimmed.length < 2) return null
  const pattern = new RegExp(`\\b${escapeRegExp(trimmed)}\\b`, 'i')
  const match = pattern.exec(text)
  return match ? match.index : null
}

function normalizeDomain(domain: string): string {
  return domain.toLowerCase().trim().replace(/^https?:\/\//, '').replace(/^www\./, '').split('/')[0] || ''
}

function isBrandName(entityName: string, brandName: string): boolean {
  const name = entityName.toLowerCase()
  const brand = brandName.toLowerCase().trim()
  if (!brand) return false
  if (name.includes(brand)) return true
  const main = brand.split(/\s+/).filter(word => word.length > 3)[0]
  return Boolean(main && main.length >= 4 && !GENERIC_WORDS.has(main) && name.includes(main))
}

function isBrandDomain(domain: string, brandDomains: string[]): boolean {
  const normalized = normalizeDomain(domain)
  if (!normalized) return false
  return brandDomains.some(candidate => {
    const brand = normalizeDomain(candidate)
    return Boolean(brand) && (normalized === brand || normalized.endsWith(`.${brand}`))
  })
}

export function findTrackedBrandPosition(
  entities: Array<Pick<FallbackEntity, 'name' | 'domain' | 'position'>>,
  brandName: string,
  brandDomains: string[] = []
): number | null {
  for (const entity of entities) {
    if (brandDomains.length > 0 && isBrandDomain(entity.domain, brandDomains)) {
      return entity.position ?? null
    }
    if (isBrandName(entity.name, brandName)) {
      return entity.position ?? null
    }
  }
  return null
}

function normalizeEntities(
  entities?: Array<EntityLike | null | undefined>
): FallbackEntity[] {
  const seen = new Set<string>()
  const normalized: FallbackEntity[] = []
  for (const entity of entities || []) {
    const name = (entity?.name || '').trim()
    if (!name) continue
    const key = name.toLowerCase()
    if (seen.has(key)) continue
    seen.add(key)
    normalized.push({
      name,
      domain: (entity?.domain || '').trim(),
      rationale: (entity?.rationale || 'Extracted from the answer text').trim(),
      position: typeof entity?.position === 'number' && entity.position > 0
        ? entity.position
        : normalized.length + 1,
    })
  }
  return normalized.sort((left, right) => left.position - right.position)
}

function knownNames(brandName: string, competitors: string[] = []): string[] {
  const names: string[] = []
  const seen = new Set<string>()
  for (const value of [brandName, ...competitors]) {
    const cleaned = (value || '').trim()
    if (!cleaned || cleaned.includes('.') || cleaned.length < 2) continue
    const key = cleaned.toLowerCase()
    if (seen.has(key)) continue
    seen.add(key)
    names.push(cleaned)
  }
  return names
}

function isUsableListName(name: string): boolean {
  const cleaned = name.trim()
  if (cleaned.length < 2 || cleaned.length > 70) return false
  if (!/[A-Za-z]/.test(cleaned)) return false
  return !/^(if you|these are|the following|note|source|http)/i.test(cleaned)
}

function extractListEntities(text: string, brandName: string, brandDomains: string[] = []): FallbackEntity[] {
  const entities: FallbackEntity[] = []
  const seen = new Set<string>()
  for (const line of text.split(/\n/)) {
    const match = line.match(/^\s*(?:#{1,6}\s*)?(?:(\d+)[.)]\s+|[-*]\s+)(?:\*\*|__)?\s*(.+)$/)
    if (!match) continue
    const explicitPosition = match[1] ? Number.parseInt(match[1], 10) : null
    let raw = (match[2] || '').replace(/\[([^\]]+)\]\([^)]+\)/g, '$1').replace(/\*\*|__|`/g, '')
    const name = raw.split(/\s*[—–|:]\s*/)[0].replace(/\s+-\s+.*$/, '').trim()
    if (!isUsableListName(name)) continue
    const key = name.toLowerCase()
    if (seen.has(key)) continue
    seen.add(key)
    entities.push({
      name,
      domain: isBrandName(name, brandName) ? (brandDomains[0] || '') : '',
      rationale: 'Extracted from a numbered or bulleted recommendation list.',
      position: explicitPosition && explicitPosition > 0 ? explicitPosition : entities.length + 1,
    })
  }
  return entities.sort((left, right) => left.position - right.position)
}

function extractNamedEntitiesFromText(input: EntityFallbackContext & { text: string }): FallbackEntity[] {
  const hits = knownNames(input.brandName, input.competitors)
    .map(name => ({ name, index: findMentionIndex(input.text, name) }))
    .filter((item): item is { name: string; index: number } => item.index !== null)
    .sort((left, right) => left.index - right.index)

  return hits.map((hit, index) => ({
    name: hit.name,
    domain: isBrandName(hit.name, input.brandName) ? (input.brandDomains?.[0] || '') : '',
    rationale: `First mentioned in the answer at character ${hit.index + 1}.`,
    position: index + 1,
  }))
}

function isUnverifiedSoloBrand(entities: FallbackEntity[], brandName: string): boolean {
  return entities.length === 1
    && isBrandName(entities[0]?.name || '', brandName)
    && (entities[0]?.position || 0) === 1
}

function extractEntitiesFromText(input: EntityFallbackContext & { text: string }): FallbackEntity[] {
  const fromList = extractListEntities(input.text, input.brandName, input.brandDomains)
  if (fromList.length >= 2) return fromList
  const fromNames = extractNamedEntitiesFromText(input)
  return fromNames.length >= 2 ? fromNames : []
}

type EntityLike = {
  name?: string | null
  domain?: string | null
  rationale?: string | null
  position?: number | null
}

export function ensureOrderedEntities(input: {
  existing?: Array<EntityLike | null | undefined>
  analysisEntities?: Array<EntityLike | null | undefined>
  brandName: string
  brandDomains?: string[]
  competitors?: string[]
  text: string
}): FallbackEntity[] {
  const context = {
    brandName: input.brandName,
    brandDomains: input.brandDomains || [],
    competitors: input.competitors || [],
    text: input.text,
  }
  const existing = normalizeEntities(input.existing)
  if (existing.length > 0 && !isUnverifiedSoloBrand(existing, input.brandName)) {
    return existing
  }
  const fromAnalysis = normalizeEntities(input.analysisEntities)
  if (fromAnalysis.length > 0 && !isUnverifiedSoloBrand(fromAnalysis, input.brandName)) {
    return fromAnalysis
  }
  return extractEntitiesFromText(context)
}

export function reconcileHallucinationFlags(
  flags: string[],
  text: string,
  brandName: string
): string[] {
  if (findMentionIndex(text, brandName) == null) {
    return [...flags]
  }
  return flags.filter(flag => flag !== 'possible_hallucination')
}

export function finalizeAnswerBlock<T extends {
  ordered_entities?: Array<EntityLike | null | undefined>
  answer_summary?: string
  notes?: { flags?: string[] | null }
}>(
  answer: T,
  context: EntityFallbackContext
): T & { ordered_entities: FallbackEntity[]; notes: { flags: string[] } } {
  const text = [context.sourceText, answer.answer_summary].filter(Boolean).join('\n')
  const orderedEntities = ensureOrderedEntities({
    existing: answer.ordered_entities,
    analysisEntities: context.analysisEntities,
    brandName: context.brandName,
    brandDomains: context.brandDomains,
    competitors: context.competitors,
    text,
  })
  const flags = reconcileHallucinationFlags(answer.notes?.flags || [], text, context.brandName)
  return {
    ...answer,
    ordered_entities: orderedEntities,
    notes: {
      ...(answer.notes || {}),
      flags,
    },
  }
}
